File: core/adaptive/adaptive_runtime_orchestrator.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from core.adaptive.adaptive_runtime_graph_engine import build_adaptive_runtime_graph
from core.adaptive.adaptive_snapshot_engine import build_adaptive_snapshot
from core.adaptive.extraction_memory_engine import remember_extraction_runtime
from core.adaptive.runtime_adaptation_engine import run_runtime_adaptation
from core.adaptive.runtime_reconciliation_engine import reconcile_runtime_state
from core.adaptive.schema_stability_engine import stabilize_extraction_schema


def run_adaptive_extraction(
    url: str,
    dom: Dict[str, Any],
    html: str,
    extraction: Dict[str, Any],
    interactions: Optional[List[Dict[str, Any]]] = None,
    memory: Optional[Dict[str, Any]] = None,
    primary_selector: str = "body",
    page: Any = None,
    stream_state: Optional[Dict[str, Any]] = None,
    identity_state: Optional[Dict[str, Any]] = None,
    pagination_state: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    dom_nodes = list(dom.get("nodes", []))
    interactions = list(interactions or [])
    memory = dict(memory or {})

    adaptation = run_runtime_adaptation(
        url=url,
        dom_nodes=dom_nodes,
        html=html,
        interactions=interactions,
        primary_selector=primary_selector,
        page=page,
    )

    schema = stabilize_extraction_schema(extraction)
    reconciliation = reconcile_runtime_state(
        browser_runtime={"available": True, "url": url},
        stream_runtime=stream_state or {},
        interaction_runtime=adaptation.get("interaction_recovery", {}),
        extraction_runtime=schema,
    )

    snapshot = build_adaptive_snapshot(
        dom=dom,
        selectors=memory.get("selectors", {}),
        interaction_state={"interactions": interactions},
        streaming_state=stream_state or {},
        identity_state=identity_state or {},
        pagination_state=pagination_state or adaptation.get("pagination_recovery", {}),
    )

    updated_memory = remember_extraction_runtime(
        memory,
        {
            "selectors": {
                primary_selector: adaptation["fallback"]["active"]["selector"],
            },
            "healed_selectors": {
                primary_selector: adaptation["fallback"]["chain"][1]["selector"],
            },
            "pagination_patterns": [
                adaptation["pagination_recovery"].get("recovered_selector", ""),
            ],
            "modal_solutions": adaptation["modal_recovery"].get("recovered", []),
            "interaction_chains": adaptation["interaction_recovery"].get("interactions", []),
        },
    )

    graph = build_adaptive_runtime_graph(adaptation)

    return {
        "adaptation": adaptation,
        "schema": schema,
        "reconciliation": reconciliation,
        "snapshot": snapshot,
        "memory": updated_memory,
        "graph": graph,
        "bounded": True,
    }

File: src/adaptive/adaptiveRuntimeOrchestrator.ts
/**
 * Converted from Python: core/adaptive/adaptive_runtime_orchestrator.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { buildAdaptiveRuntimeGraph } from "./adaptiveRuntimeGraphEngine.js";
import { buildAdaptiveSnapshot } from "./adaptiveSnapshotEngine.js";
import { rememberExtractionRuntime } from "./extractionMemoryEngine.js";
import { runRuntimeAdaptation } from "./runtimeAdaptationEngine.js";
import { reconcileRuntimeState } from "./runtimeReconciliationEngine.js";
import { stabilizeExtractionSchema } from "./schemaStabilityEngine.js";

export function runAdaptiveExtraction(url: any, dom: any, html: any, extraction: any, interactions: any = null, memory: any = null, primary_selector: any = "body", page: any = null, stream_state: any = null, identity_state: any = null, pagination_state: any = null): any {
  var dom_nodes: any = [...py.iter(py.get(dom, "nodes", []))];
  interactions = [...py.iter(py.or2(interactions, () => ([])))];
  memory = py.pyDict(py.or2(memory, () => ({})));
  var adaptation: any = runRuntimeAdaptation(url, dom_nodes, html, interactions, primary_selector, page);
  var schema: any = stabilizeExtractionSchema(extraction);
  var reconciliation: any = reconcileRuntimeState({"available": true, "url": url}, py.or2(stream_state, () => ({})), py.get(adaptation, "interaction_recovery", {}), schema);
  var snapshot: any = buildAdaptiveSnapshot(dom, py.get(memory, "selectors", {}), {"interactions": interactions}, py.or2(stream_state, () => ({})), py.or2(identity_state, () => ({})), py.or2(pagination_state, () => (py.get(adaptation, "pagination_recovery", {}))));
  var updated_memory: any = rememberExtractionRuntime(memory, {"selectors": {[py.toStr(primary_selector)]: py.at(py.at(py.at(adaptation, "fallback"), "active"), "selector")}, "healed_selectors": {[py.toStr(primary_selector)]: py.at(py.at(py.at(py.at(adaptation, "fallback"), "chain"), 1), "selector")}, "pagination_patterns": [py.get(py.at(adaptation, "pagination_recovery"), "recovered_selector", "")], "modal_solutions": py.get(py.at(adaptation, "modal_recovery"), "recovered", []), "interaction_chains": py.get(py.at(adaptation, "interaction_recovery"), "interactions", [])});
  var graph: any = buildAdaptiveRuntimeGraph(adaptation);
  return {"adaptation": adaptation, "schema": schema, "reconciliation": reconciliation, "snapshot": snapshot, "memory": updated_memory, "graph": graph, "bounded": true};
}
export { buildAdaptiveRuntimeGraph, buildAdaptiveSnapshot, reconcileRuntimeState, rememberExtractionRuntime, runRuntimeAdaptation, stabilizeExtractionSchema };
